Source/Core/DolphinQt2/CheatsManager.cpp

```diff
@@ -0,0 +1,633 @@
+          &CheatsManager::OnWatchContextMenu);
+  connect(m_watch_table, &QTableWidget::itemChanged, this, &CheatsManager::OnWatchItemChanged);
+}
+
+void CheatsManager::OnWatchContextMenu()
+{
+  if (m_watch_table->selectedItems().isEmpty())
+    return;
+
+  QMenu* menu = new QMenu(this);
+
+  AddAction(menu, tr("Remove from Watch"), this, [this] {
+    auto* item = m_match_table->selectedItems()[0];
+
+    int index = item->data(INDEX_ROLE).toInt();
+
+    m_watch.erase(m_watch.begin() + index);
+
+    Update();
+  });
+
+  menu->addSeparator();
+
+  AddAction(menu, tr("Generate Action Replay Code"), this, &CheatsManager::GenerateARCode);
+
+  menu->exec(QCursor::pos());
+}
+
+void CheatsManager::OnMatchContextMenu()
+{
+  if (m_match_table->selectedItems().isEmpty())
+    return;
+
+  QMenu* menu = new QMenu(this);
+
+  AddAction(menu, tr("Add to Watch"), this, [this] {
+    auto* item = m_match_table->selectedItems()[0];
+
+    int index = item->data(INDEX_ROLE).toInt();
+
+    m_watch.push_back(m_results[index]);
+
+    Update();
+  });
+
+  menu->exec(QCursor::pos());
+}
+
+static ActionReplay::AREntry ResultToAREntry(Result result)
+{
+  u8 cmd;
+
+  switch (result.type)
+  {
+  case DataType::Byte:
+    cmd = 0x00;
+    break;
+  case DataType::Short:
+    cmd = 0x02;
+    break;
+  default:
+  case DataType::Int:
+    cmd = 0x04;
+    break;
+  }
+
+  u32 address = result.address & 0xffffff;
+
+  return ActionReplay::AREntry(cmd << 24 | address, result.locked_value);
+}
+
+void CheatsManager::GenerateARCode()
+{
+  if (!m_ar_code)
+    return;
+
+  auto* item = m_match_table->selectedItems()[0];
+
+  int index = item->data(INDEX_ROLE).toInt();
+  ActionReplay::ARCode ar_code;
+
+  ar_code.active = true;
+  ar_code.user_defined = true;
+  ar_code.name = tr("Generated by search (Address %1)")
+                     .arg(m_watch[index].address, 8, 16, QLatin1Char('0'))
+                     .toStdString();
+
+  ar_code.ops.push_back(ResultToAREntry(m_watch[index]));
+
+  m_ar_code->AddCode(ar_code);
+}
+
+void CheatsManager::OnWatchItemChanged(QTableWidgetItem* item)
+{
+  if (m_updating)
+    return;
+
+  int index = item->data(INDEX_ROLE).toInt();
+  int column = item->data(COLUMN_ROLE).toInt();
+
+  switch (column)
+  {
+  case 0:
+    m_watch[index].name = item->text();
+    break;
+  case 3:
+    m_watch[index].locked = item->checkState() == Qt::Checked;
+    break;
+  case 4:
+  {
+    const auto text = item->text();
+    u64 value = 0;
+
+    switch (static_cast<DataType>(m_match_length->currentIndex()))
+    {
+    case DataType::Byte:
+      value = text.toUShort(nullptr, 16) & 0xFF;
+      break;
+    case DataType::Short:
+      value = text.toUShort(nullptr, 16);
+      break;
+    case DataType::Int:
+      value = text.toUInt(nullptr, 16);
+      break;
+    case DataType::Float:
+    {
+      float f = text.toFloat();
+      std::memcpy(&value, &f, sizeof(float));
```

Comment:
Is that safe with `f` and `value` having different size followed by the 32-bit truncation by `locked_value`? I'm pretty sure it works because it's little endian. Otherwise, I think it would break on big endian architecture due to the low 32-bit being on the other side.

Reply:
Not applicable anymore.